### ISSUE TITLE
Memory leak, possibly AudioContainer

### ISSUE THREAD

**Issue by RicherMans** [turpaultn/DCASE2019_task4#3]
Hey guys,
as with last years script, I can't download the dataset, since this script uses more and more memory while downloading, independently of the number of threads. 
I got a 16GB RAM machine here and the download script uses at the beginning ~2% and goes up ( after 300 utterances ) to ~10% and later just crashes the computer.
The memory fills slowly up when downloading, meaning that in order to download the dataset I need to run the script many times over.
It looks like the culprit is `AudioConainer`. For each download you create an `AudioContainer` object. It could be that internally some reference within `AudioContainer` is never freed, therefore memory never released. Moreover, the use of this object in this context ( just for copying and resampling a file ) is maybe a bit overkill. A simple change to `librosa.load` / `librosa.output.write_wav` solves the entire problem.

Other things that  I noticed along the way in order to resolve this problem are:

1. Missing files are (with errors) kept in memory up until everything finishes. This shouldn't be a big deal, but could simply fill up memory if too many errors happen.
2. Returning lists, even though they are never used again. I would recommend using tuples, saves quite some memory for a lot of data.

Can i submit a pull request (change to librosa) for this issue?

EDIT:
Its not `AudioContainer` or any variable within the `download` function. After using memory_profiler for some time, I still dont have a clue. I suggest maybe some of the logs is filled up.

**Comment by turpaultn** [comment]
Hi,
I'm trying to reproduce but without success.
I tried the code on macOS High Sierra (8GB RAM) and ubuntu 18.04 (32GB RAM).
I created a new environment in both cases (thanks to the script in the repo).

I am computing with N_JOBS=3 and CHUNK_SIZE=10 (default values).
I have 3 processes turning in parallel, as expected and each of them has a variable size going from 90MB to 170MB. 
Indeed, we're always storing the full youtube file in RAM, so it takes some space.

However, the global amount of memory is very stable in both cases.
(I would avoid the librosa pull request since the container is checking the format in a good way and apply the good transformation. I tried to do it without AudioContainer, but I had more or less have to do the same kind of things.)

**Comment by turpaultn** [comment]
Sorry, I didn't see your edit.
You can change the logger level in utils/Logger.py if you want.
Do not hesitate to update if you find find the issue.

**Comment by RicherMans** [comment]
Hey thanks!
I'm still searching for the reason why my download is horribly overusing memory.
I agree it is not anything directly related to your script. I use arch linux with python3.7.2.

Is it some OS related stuff that you might think of? I check to use some Ubunutu or something nearby, thanks!

**Comment by turpaultn** [comment]
I don't know, but I would think it can come from the multiprocessing.
Do you still have the same problem when with N_JOBS=1 ? (to see if it could come from the multiprocessing module)

**Comment by RicherMans** [comment]
Yup same problem no matter how many threads or if N_JOBS=1.
I tried to free the variables in download (meta and audiocontainer) before returning in order to verify if there is somewhere a pendling variable.
Currently I think it would be something related to my shadowsocks proxy, i"ll try to use a vpn directly and let you know. 
Thanks!

**Comment by turpaultn** [comment]
Ok, don't hesitate to update, we'll try to help.
For your information: for other participants using a VPN, the --proxy option of youtube-dl worked and nobody reported other errors.

**Comment by RicherMans** [comment]
Hey there,
so after couple of days of checking everything, from the proxychains tool ( there are 3 available ) to the possibility that it might be directly my local proxy ( the server ) or even using a vpn, all lead to the same result : Memory overflow
After checking with my labmate on his ubuntu machine ( old python versions yay! ), I noticed that python 3.6.7 has no trouble downloading.
On my machines ( both arch ), I use python 3.7.2 ... there is something with these versions which lead to this memory leak. Can't tell why, but next time I'll just go with a conda environment.

TLDR; Use python3.6 not 3.7

Thanks guys!
